### ISSUE TITLE
 tensor_info.GetShape() gives [-1, 1 ] as shape.

### ISSUE THREAD

**Issue by Karnav123** [microsoft/onnxruntime#4051]
**Describe the bug**
I have an xgboost model in onnx format trained on titanic dataset from Kaggle with 5 input nodes. When I load the model in onnx runtime using C++ API, the shape of the  input node comes out to be [-1, 1].

**Urgency**
I need to use this C++ API in a production code.

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows  1909
- ONNX Runtime installed from (source or binary): Nuget Package in VS2019
- ONNX Runtime version: 1.2.0
- Python version: 3.7
- Visual Studio version (if applicable): 2019
- GCC/Compiler version (if compiling from source): MSVC
- CUDA/cuDNN version: NA
- GPU model and memory: NA

**To Reproduce**
- Describe steps/code to reproduce the behavior.
```
int main(int argc, char* argv[]) {
    //*************************************************************************
    // initialize  enviroment...one enviroment per process
    // enviroment maintains thread pools and other state info
    Ort::Env env(ORT_LOGGING_LEVEL_WARNING, "test");

    // initialize session options if needed
    Ort::SessionOptions session_options;
    session_options.SetIntraOpNumThreads(1);

    // If onnxruntime.dll is built with CUDA enabled, we can uncomment out this line to use CUDA for this
    // session (we also need to include cuda_provider_factory.h above which defines it)
    // #include "cuda_provider_factory.h"
    // OrtSessionOptionsAppendExecutionProvider_CUDA(session_options, 1);

    // Sets graph optimization level
    // Available levels are
    // ORT_DISABLE_ALL -> To disable all optimizations
    // ORT_ENABLE_BASIC -> To enable basic optimizations (Such as redundant node removals)
    // ORT_ENABLE_EXTENDED -> To enable extended optimizations (Includes level 1 + more complex optimizations like node fusions)
    // ORT_ENABLE_ALL -> To Enable All possible opitmizations
    session_options.SetGraphOptimizationLevel(GraphOptimizationLevel::ORT_ENABLE_EXTENDED);

    //*************************************************************************
    // create session and load model into memory
    // using squeezenet version 1.3
    // URL = https://github.com/onnx/models/tree/master/squeezenet
#ifdef _WIN32
    const wchar_t* model_path = L"pipeline_titanic.onnx";
#else
    const char* model_path = "squeezenet.onnx";
#endif

    printf("Using Onnxruntime C++ API\n");
    Ort::Session session(env, model_path, session_options);

    //*************************************************************************
    // print model input layer (node names, types, shape etc.)
    Ort::AllocatorWithDefaultOptions allocator;

    // print number of model input nodes
    size_t num_input_nodes = session.GetInputCount();
    std::vector<const char*> input_node_names(num_input_nodes);
    std::vector<std::vector<int64_t>> input_node_dims(num_input_nodes);  // simplify... this model has only 1 input node {1, 3, 224, 224}.
                                           // Otherwise need vector<vector<>>

    printf("Number of inputs = %zu\n", num_input_nodes);

    // iterate over all input nodes
    for (int i = 0; i < num_input_nodes; i++) {
        // print input node names
        char* input_name = session.GetInputName(i, allocator);
        printf("Input %d : name=%s\n", i, input_name);
        input_node_names[i] = input_name;

        // print input node types
        Ort::TypeInfo type_info = session.GetInputTypeInfo(i);
        auto tensor_info = type_info.GetTensorTypeAndShapeInfo();

        ONNXTensorElementDataType type = tensor_info.GetElementType();
        printf("Input %d : type=%d\n", i, type);

        // print input shapes/dims
        input_node_dims[i] = tensor_info.GetShape();
        printf("Input %d : num_dims=%zu\n", i, input_node_dims.size());
        for (int j = 0; j < input_node_dims[i].size(); j++)
            printf("Input %d : dim %d=%jd\n", i, j, input_node_dims[i][j]);
    }
```

The output is;
Input 0 : type=8
Input 0 : num_dims=5
Input 0 : dim 0=-1
Input 0 : dim 1=1
Input 1 : name=sex
Input 1 : type=8
Input 1 : num_dims=5
Input 1 : dim 0=-1
Input 1 : dim 1=1
Input 2 : name=age
Input 2 : type=1
Input 2 : num_dims=5
Input 2 : dim 0=-1
Input 2 : dim 1=1
Input 3 : name=fare
Input 3 : type=1
Input 3 : num_dims=5
Input 3 : dim 0=-1
Input 3 : dim 1=1
Input 4 : name=embarked
Input 4 : type=8
Input 4 : num_dims=5
Input 4 : **dim 0=-1**    // How dim is -1
Input 4 : dim 1=1

- Attach the ONNX model to the issue (where applicable) to expedite investigation.
Download the model from the following link.
[(https://github.com/Karnav123/onnx_model/blob/master/pipeline_titanic.onnx)]

**Expected behavior**
The dimension of the input node should not be negative.

**Screenshots**

**Additional context**
Add any other context about the problem here. If the issue is about a particular model, please share the model details as well to facilitate debugging.

**Comment by snnn** [comment]
Here "-1" means "it can be **any** non-negative number."

**Comment by Karnav123** [comment]
> Here "-1" means "it can be **any** non-negative number."

OK. But if I pass this shape in CreateTensor it gives an error. Moreover, in this model there are multiple input nodes so how shall I consider the shape of the input data?

**Comment by Craigacp** [comment]
The first dimension is the batch size. So set it to however big the batch is at inference time. You'll need to supply 5 input tensors all size [batch_size,1], each containing the feature values for that specific feature. It's a bit weird that the xgboost converter didn't make it a single input of size [batch_size,5]. Maybe there are some options you could tweak when you convert the model?

**Comment by Karnav123** [comment]
> The first dimension is the batch size. So set it to however big the batch is at inference time. You'll need to supply 5 input tensors all size [batch_size,1], each containing the feature values for that specific feature. It's a bit weird that the xgboost converter didn't make it a single input of size [batch_size,5]. Maybe there are some options you could tweak when you convert the model?

Currently, I am doing the following. I have hardcoded the data for evaluation.
`size_t input_tensor_size = 2;  // simplify ... using known dim values to calculate size
                                               // use OrtGetTensorShapeElementCount() to get official size!

    /*pclass
        sex
        age
        fare
        embarked*/
    std::vector<std::string> input_tensor_values_pclass = {"1", "3"};
    std::vector<std::string> input_tensor_values_sex = { "male" , "female"};
    std::vector<float> input_tensor_values_age = { 25.0, 34.0 };
    std::vector<float> input_tensor_values_fare = { 7.4958, 55.9};
    std::vector<std::string> input_tensor_values_embarked = { "S", "Q" };

    std::vector<int64_t> input_dim{ 0, 1 };
    //std::vector<const char*> output_node_names = { "softmaxout_1" };

    // initialize input data with values in [0.0, 1.0]
    /*for (unsigned int i = 0; i < input_tensor_size; i++)
        input_tensor_values[i] = (float)i / (input_tensor_size + 1);*/

    // create input tensor object from data values
    auto memory_info = Ort::MemoryInfo::CreateCpu(OrtArenaAllocator, OrtMemTypeDefault);
    Ort::Value input_tensor_pclass = Ort::Value::CreateTensor(memory_info, input_tensor_values_pclass.data(), input_tensor_size, input_dim.data(), 2, ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_STRING);
    assert(input_tensor_pclass.IsTensor());

    Ort::Value input_tensor_sex = Ort::Value::CreateTensor(memory_info, input_tensor_values_sex.data(), input_tensor_size, input_dim.data(), 2, ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_STRING);
    assert(input_tensor_sex.IsTensor());

    Ort::Value input_tensor_age = Ort::Value::CreateTensor<float>(memory_info, input_tensor_values_age.data(), input_tensor_size, input_dim.data(), 2);
    assert(input_tensor_age.IsTensor());

    Ort::Value input_tensor_fare = Ort::Value::CreateTensor<float>(memory_info, input_tensor_values_fare.data(), input_tensor_size, input_dim.data(), 2);
    assert(input_tensor_fare.IsTensor());

    Ort::Value input_tensor_embarked = Ort::Value::CreateTensor(memory_info, input_tensor_values_embarked.data(), input_tensor_size, input_dim.data(), 2, ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_STRING);
    assert(input_tensor_embarked.IsTensor());`

Here, if I change the 0 to any other value it gives an error. I am not able to proceed from here.

**Comment by prabhat00155** [comment]
Look at this: https://github.com/microsoft/onnxruntime/issues/3986#issuecomment-635894186

**Comment by prabhat00155** [comment]
Closing this as the question already answered in the other thread linked above.
